crates/agent/src/state.rs

```diff
@@ -76,49 +90,148 @@ impl GlobalState {
-            let timeout = tokio::time::sleep(NODE_GRACEFUL_SHUTDOWN_TIMEOUT);
-
-            select! {
-                _ = child.wait() => (),
-                _ = timeout => {
-                    info!("snarkos process did not gracefully shut down, killing...");
-                    child.kill().await.unwrap();
+    pub async fn update_agent_state(&self, state: AgentState, opts: ReconcileOptions) {
+        if let Err(e) = self.db.set_agent_state(&state) {
+            error!("failed to save agent state to db: {e}");
+        }
+        let state = Arc::new(state);
+        *self.agent_state.write().await = state;
+
+        // Queue a reconcile to apply the new state
+        self.queue_reconcile(Duration::ZERO, opts).await;
+    }
+
+    pub async fn re_fetch_peer_addrs(&self) {
+        let agent_state = self.get_agent_state().await;
+        let AgentState::Node(_, node) = agent_state.as_ref() else {
+            return;
+        };
+
+        let Some(client) = self.get_ws_client().await else {
+            return;
+        };
+
+        let peer_ids = node
+            .peers
+            .iter()
+            .chain(node.validators.iter())
+            .filter_map(|p| {
+                if let snops_common::state::AgentPeer::Internal(id, _) = p {
+                    Some(*id)
+                } else {
+                    None
                 }
+            })
+            // Ensure we only have unique agent ids (can use itertools down the line)
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect::<Vec<_>>();
+
+        if peer_ids.is_empty() {
+            return;
+        }
+
+        let new_addrs = match client.resolve_addrs(context::current(), peer_ids).await {
+            Ok(Ok(new_addrs)) => new_addrs,
+            Ok(Err(e)) => {
+                error!("Control plane failed to resolve addresses: {e}");
+                return;
+            }
+            Err(e) => {
+                error!("RPC failed to resolve addresses: {e}");
+                return;
             }
+        };
+
+        // Extend the cache with the updated addrs
+        let mut lock = self.resolved_addrs.write().await;
+        let has_new_addr = new_addrs
+            .iter()
+            .any(|(id, addr)| lock.get(id) != Some(addr));
+
+        if !has_new_addr {
+            return;
+        }
+
+        info!("Resolved updated addrs from handshake");
+
+        lock.extend(new_addrs);
+        if let Err(e) = self.db.set_resolved_addrs(Some(&lock)) {
+            error!("failed to save resolved addrs to db: {e}");
         }
     }
 
-    pub fn transfer_tx(&self) -> TransferTx {
-        self.transfer_tx.clone()
+    pub async fn set_node_status(&self, status: Option<SnarkOSStatus>) {
+        *self.last_node_status.write().await = status.map(|s| (Instant::now(), s));
+    }
+
+    pub async fn get_node_status(&self) -> Option<SnarkOSStatus> {
```

Comment:
Couldn't this be a reference? Or is needed as a non-reference where it's called?

Reply:
I think it would have to be the RwLock's guard type, which felt kinda weird. Any references would be temporary